python/cudf/udf_cpp/shim.cu

```diff
@@ -662,3 +664,41 @@ make_definition_idx(BlockIdxMax, int64, int64_t);
 make_definition_idx(BlockIdxMax, float64, double);
 #undef make_definition_idx
 }
+
+__device__ void udf_str_dtor(void* udf_str, size_t size, void* dtor_info)
+{
+  udf_string* ptr = reinterpret_cast<udf_string*>(udf_str);
+  ptr->~udf_string();
+}
+
+// Only used to allocate the right amount of space, see below
+struct meminfo_and_str {
+  NRT_MemInfo mi;
+  udf_string st;
+};
+
+/*
+Create a new MemInfo object holding the reference count of a udf_string. When returning
+new strings, shim functions expect a pointer to a stack allocated buffer into which it
+will construct the udf_string it returns. Since one can not safely build a MemInfo object
+around this stack memory, we store a copy of the udf_string itself next to the MemInfo
+where it can be later used to destroy the underlying data.
+*/
+extern "C" __device__ int meminfo_from_new_udf_str(void** nb_retval, void* udf_str)
+{
+  // allocate enough room for both the meminfo and udf_string
+  meminfo_and_str* mi_and_str = (meminfo_and_str*)NRT_Allocate(sizeof(meminfo_and_str));
+  if (mi_and_str != NULL) {
+    auto mi_ptr        = &(mi_and_str->mi);
+    udf_string* st_ptr = &(mi_and_str->st);
+
+    NRT_MemInfo_init(mi_ptr, st_ptr, NULL, udf_str_dtor, NULL);
```

Comment:
It's a bit confusing for the uninitiated reader to follow seeing the size being null - perhaps:
```suggestion
    // We pass a null size here because the udf_string actually exists on the heap, 
    // and tracks the size of the string data that it points to.
    NRT_MemInfo_init(mi_ptr, st_ptr, NULL, udf_str_dtor, NULL);
```

Reply:
I think I wrote "heap" when I meant "stack" there, apologies for confusion